enterprise/reporting/mediator.go

```diff
@@ -13,77 +13,66 @@ import (
 )
 
 type ReportingMediator struct {
-	log             logger.Logger
-	once            sync.Once
-	enterpriseToken string
-	reporting       types.Reporting
-	errorReporting  types.Reporting
+	log logger.Logger
+
+	delegates []types.Reporting
 }
 
-func (med *ReportingMediator) createReportInstance(backendConfig backendconfig.BackendConfig) types.Reporting {
-	med.log.Debug("Forming reporting instance")
-	reportingEnabled := config.GetBool("Reporting.enabled", types.DefaultReportingEnabled)
-	if !reportingEnabled || med.enterpriseToken == "" {
-		return &NOOP{}
+func NewReportingMediator(ctx context.Context, log logger.Logger, enterpriseToken string, backendConfig backendconfig.BackendConfig) *ReportingMediator {
```

Comment:
Here we are passing ctx, but the `NewReportingMediator` method is not really blocking. It runs go routines in the background without a way to block until they return. 



Would it make sense to use the `DatabaseSyncer` as a generic `Run` method, that will accept ctx and will be responsible for running all the `defaultReporter.backendConfigSubscriber` go routines? 

And simplify  NewReportingMediator:
```suggestion
func NewReportingMediator(log logger.Logger, enterpriseToken string, backendConfig backendconfig.BackendConfig) *ReportingMediator {
```

to just create the  ReportingMediator struct:
```go
type ReportingMediator struct {
	log logger.Logger

	delegates []types.Reporting
	backconfSubscribers []func(ctx, ...) 
 }
 ```go
 
 Then `DatabaseSyncer` will use its waitgroup to run delegates and backend config subscribers

Reply:
Blocking until all go routines have stopped is not implemented yet. There are 2 different kinds of go routines:

1. backend config subscribers
2. syncers' main loops

Both would need to stop during shutdown. The current improvement is that go routines will stop when the context is canceled, even if the program currently doesn't wait for them to stop.

Reply:
will try to take care of it in this pr

Reply:
@lvrach reporting mediator is now using its own `errgroup`, it spawns goroutines using this group and waits for all of them to finish before returning from its `ReportingSyncer` functions.
`gateway`, `processor` and `warehouse` are all running `ReportingSyncer`s in their own errgroups, i.e. waiting for them to finish.